src/crimson/os/seastore/extent_placement_manager.cc

```diff
@@ -473,6 +474,9 @@ ExtentPlacementManager::BackgroundProcess::try_reserve(
     if (res.reserve_ool_success) {
       major_cleaner->release_projected_usage(usage.inline_usage + usage.ool_usage);
     }
+    if (cold_cleaner && res.reserve_cold_ool_success) {
+      cold_cleaner->release_projected_usage(usage.cold_ool_usage);
+    }
   }
   return res;
 }
```

Comment:
The overall logic looks reasonable to me, how about making the reservations encapsualted and reuse them where possible, probably with the below helpers?

```c++
struct cleaner_usage_t {
  std::size_t main_usage = 0;
  std::size_t cold_ool_usage = 0;
};

struct io_usage_t {
  std::size_t inline_usage = 0;
  cleaner_usage_t cleaner_usage;
};

struct reserve_cleaner_result_t {
   bool reserve_main_success = true;
   bool reserve_cold_ool_success = true;

   bool is_successful() const {
     return reserve_main_ool_success &&
            reserve_cold_ool_success;
   }
};

struct reserve_io_result_t {
   bool reserve_inline_success = true;
   reserve_cleaner_result_t reserve_cleaner;

   bool is_successful() const {
     return reserve_inline_success &&
            reserve_cleaner.is_successful();
   }
};

/*
 * reusable BackgroundProcess methods
 */

void BackgroundProcess::abort_cold_usage(std::size_t cold_ool_usage) {
  assert(is_ready());
  if (has_cold_tier()) {
    cold_cleaner->release_projected_usage(cold_ool_usage);
  } else {
    assert(cold_ool_usage == 0);
  }
}

bool BackgroundProcess::try_reserve_cold(std::size_t cold_ool_usage) {
  assert(is_ready());
  if (has_cold_tier()) {
    return cold_cleaner->try_reserve_projected_usage(cold_ool_usage);
  } else {
    assert(cold_ool_usage == 0);
    return true;
  }
}

void BackgroundProcess::abort_cleaner_usage(
    const reserve_cleaner_result_t &res,
    const cleaner_usage_t &usage) {
  assert(is_ready());
  if (res.reserve_main_success) {
    main_cleaner->release_projected_usage(usage.main_usage);
  }
  if (res.reserve_cold_ool_success) {
    abort_cold_usage(usage.cold_ool_usage);
  }
}

reserve_cleaner_result_t BackgroundProcess::try_reserve_cleaner(const cleaner_usage_t &usage) {
  assert(is_ready());
  reserve_cleaner_result_t res {
    major_cleaner->try_reserve_projected_usage(usage.main_usage),
    try_reserve_cold(usage.cold_ool_usage)
  };
  if (!res.is_successful()) {
    abort_cleaner_usage(res, usage);
  }
}

void BackgroundProcess::abort_io_usage(
    const reserve_io_result_t &res,
    const cleaner_io_usage_t &usage) {
  assert(is_ready());
  if (res.reserve_inline_success) {
    trimmer->release_inline_usage(usage.inline_usage);
  }
  abort_cleaner_usage(res.reserve_cleaner, usage.cleaner_usage);
}

reserve_io_result_t BackgroundProcess::try_reserve_io(const io_usage_t &usage) {
  assert(is_ready());
  reserve_io_result_t res {
    trimmer->try_reserve_inline_usage(usage.inline_usage),
    try_reserve_cleaner(usage.cleaner_usage)
  }
  if (!res.is_successful()) {
    abort_io_usage(res, usage);
  }
}
```